File: python/fetch_announcements.py
from flask import Flask, jsonify
import requests
import time
import os
import logging

app = Flask(__name__)
LAST_CHECK_FILE = "last_check.txt"
DATA_URL = "https://raw.githubusercontent.com/Sekai-World/sekai-master-db-diff/main/userInformations.json"

# loggingの設定
logging.basicConfig(level=logging.INFO)  # ログレベルをINFOにして冗長なログを減らす

# 前回のチェック時刻を取得
def get_last_check_time():
    if os.path.exists(LAST_CHECK_FILE):
        with open(LAST_CHECK_FILE, "r") as f:
            return int(f.read().strip())
    
    # 初回実行時は現在の時刻を保存して、通知は送らない
    current_time = int(time.time())
    save_last_check_time(current_time)
    return current_time  

# 現在のチェック時刻を保存
def save_last_check_time(timestamp):
    with open(LAST_CHECK_FILE, "w") as f:
        f.write(str(timestamp))

@app.route("/announcements", methods=["GET"])
def fetch_announcements():
    last_check = get_last_check_time()
    current_time = int(time.time())  # 現在のUNIX時刻
    logging.info(f"Checking announcements from {last_check} to {current_time}")

    try:
        response = requests.get(DATA_URL)
        response.raise_for_status()
        announcements = response.json()
    except Exception as e:
        logging.error(f"API Fetch Error: {e}")
        return f"エラー: {str(e)}", 500

    new_announcements = []

    for item in announcements:
        start_at_seconds = item["startAt"] // 1000  # ミリ秒を秒に変換
        if last_check <= start_at_seconds <= current_time:  # last_check から現在の時刻の範囲のみ
            if item["browseType"] == "internal":
                new_announcements.append(item["title"])
            elif item["browseType"] == "external":
                new_announcements.append(f'[{item["title"]}]({item["path"]})')

    if new_announcements:
        save_last_check_time(current_time)  # 通知があった場合のみ更新
        return "\n".join(new_announcements)
    
    return "新しいお知らせはありません。"

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000)

File: announce-fetcher/fetch_announcements.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"
)

const (
	lastCheckFile = "last_check.txt"
	dataURL       = "https://raw.githubusercontent.com/Sekai-World/sekai-master-db-diff/main/userInformations.json"
)

type Announcement struct {
	Title      string `json:"title"`
	StartAt    int64  `json:"startAt"`
	BrowseType string `json:"browseType"`
	Path       string `json:"path"`
}

func getLastCheckTime() int64 {
	if _, err := os.Stat(lastCheckFile); err == nil {
		data, err := ioutil.ReadFile(lastCheckFile)
		if err == nil {
			var t int64
			fmt.Sscanf(string(data), "%d", &t)
			return t
		}
	}

	now := time.Now().Unix()
	saveLastCheckTime(now)
	return now
}

func saveLastCheckTime(t int64) {
	ioutil.WriteFile(lastCheckFile, []byte(fmt.Sprintf("%d", t)), 0644)
}

func fetchAnnouncements(w http.ResponseWriter, r *http.Request) {
	lastCheck := getLastCheckTime()
	currentTime := time.Now().Unix()

	resp, err := http.Get(dataURL)
	if err != nil {
		http.Error(w, fmt.Sprintf("エラー: %v", err), http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		http.Error(w, fmt.Sprintf("エラー: %v", err), http.StatusInternalServerError)
		return
	}

	var announcements []Announcement
	if err := json.Unmarshal(body, &announcements); err != nil {
		http.Error(w, fmt.Sprintf("エラー: %v", err), http.StatusInternalServerError)
		return
	}

	var newAnnouncements []string

	for _, item := range announcements {
		startAtSeconds := item.StartAt / 1000

		if startAtSeconds >= lastCheck && startAtSeconds <= currentTime {
			if item.BrowseType == "internal" {
				newAnnouncements = append(newAnnouncements, item.Title)
			} else if item.BrowseType == "external" {
				newAnnouncements = append(newAnnouncements, fmt.Sprintf("[%s](%s)", item.Title, item.Path))
			}
		}
	}

	if len(newAnnouncements) > 0 {
		saveLastCheckTime(currentTime)
		for _, a := range newAnnouncements {
			fmt.Fprintln(w, a)
		}
		return
	}

	fmt.Fprintln(w, "新しいお知らせはありません。")
}

func main() {
	http.HandleFunc("/announcements", fetchAnnouncements)
	log.Fatal(http.ListenAndServe(":5000", nil))
}